### ISSUE TITLE
Symtab reads wrong debug inline info for Intel EM_INTELGT GPU binary

### ISSUE THREAD

**Issue by mwkrentel** [dyninst/dyninst#1469]
**Describe the bug**
I'm seeing a problem with symtabAPI reading the dwarf inline info from
a new Intel GPU binary.  The problem is that the binary has two File
Name Tables and symtab is looking up the file name index from the
dwarf DIE in the wrong table, hence it gets the wrong file name.
Note: this is the inline sequence, not the terminal line map.

The binary (`03fff22c.gpubin`) is from sunspot at ALCF, but sunspot is
currently down for extended maintenance, so I'm running everything on
ufront.cs at Rice.

```
$ file 03fff22c.gpubin 
03fff22c.gpubin: ELF 64-bit LSB executable, *unknown arch 0xcd* version 1 (SYSV),
statically linked, with debug_info, not stripped
```

This is using Dyninst master as of July 27, 2023 (8fa753d955d4) which
includes JohnMC's PR to add binary type EM_INTELGT.  Note that 0xcd =
205 = EM_INTELGT.

I dug out my old `openmp-symtab.cpp` program from:

```
https://github.com/hpctoolkit/hpctoolkit-tests.git
cd hpctoolkit-tests/dyninst
```

This just opens a file, looks up every symtab function, and queries
the line map and inline info every 4 bytes.  Doesn't touch parseAPI or
instructionAPI.

Look at the inline info for 0x8000ffe800f8.

```
stmt:  0x8000ffe800f8  l=87  f='/home/johnmc/va/src/vector-add-buffers.cpp'
   inline:  l=1247
   f='/soft/restricted/CNDA/updates/2023.05.15.001/oneapi/compiler/eng-20230512/compiler/linux/bin-llvm/../include/sycl/item.hpp'
   p='_ZN4sycl3_V18accessorIiLi1ELNS0_6access4modeE1024ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_3ext6oneapi22accessor_property_listIJEEEE6__initEPU3AS1iNS0_5rangeILi1EEESE_NS0_2idILi1EEE'
```

That's line 1247 from file item.hpp.  Thing is, item.hpp only has 167
lines, not 1247.  So, where does item.hpp come from?
Look at one of the DIEs:

```
 <2><f594>: Abbrev Number: 4 (DW_TAG_inlined_subroutine)
    <f595>   DW_AT_abstract_origin: <0x1e01a>
    <f599>   DW_AT_call_file   : 28
    <f59a>   DW_AT_call_line   : 1247
    <f59c>   DW_AT_ranges      : 0x160
```

That's line 1247 from file index 28.  Now we check the File Name
Table.  Actually, there are TWO File Name Tables.

```
 The File Name Table (offset 0x309):
  Entry Dir     Time    Size    Name
  1     0       0       0       vector-add-buffers.cpp
  2     1       0       0       handler.hpp
  ...
  27    4       0       0       item_base.hpp
  28    1       0       0       item.hpp

 The File Name Table (offset 0x889):
  Entry Dir     Time    Size    Name
  1     0       0       0       vector-add-buffers.cpp
  2     1       0       0       access.hpp
  ...
  27    2       0       0       item.hpp
  28    2       0       0       handler.hpp
```

So, in one table, 28 is handler.hpp (correct), but in the other, 28 is
item.hpp (incorrect).  For some reason, it appears that symtab is
looking up file #28 in the wrong table.

Note: having multiple File Name Tables by itself is not a problem.
You probably get one per comp unit.  The hpcstruct binary has 42 file
tables and symtab reads that Ok.

However 03fff22c.gpubin has two .text segments (hpcstruct-bin has only
one).  Maybe that's it, I don't know.

```
Section Headers:
  [Nr] Name              Type            Address          Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            0000000000000000 000000 000000 00      0   0  8
  [ 1] .text._ZTSN4sycl3_V16detail18RoundedRangeKernelINS0_4itemILi1ELb1EEELi1EZZ9VectorAddRNS0_5queueERKSt6vectorIiSaIiEESB_RS9_ENKUlRNS0_7handlerEE_clESE_EUlT_E_EE PROGBITS        00008000ffe90000 0005f0 000300 00  AX  0   0  8
  [ 2] .text._ZTSZZ9VectorAddRN4sycl3_V15queueERKSt6vectorIiSaIiEES7_RS5_ENKUlRNS0_7handlerEE_clESA_EUlT_E_ PROGBITS        00008000ffe80000 0008f0 0002c0 00  AX  0   0  8
  [ 3] .symtab           SYMTAB          0000000000000000 000bb0 000150 18     20  13  8
```

**System (please complete the following information):**
This happened on sunspot at ANL which is down for extended maintenance.

I have all the relevant files (gpu binary, unit test, output of
objdump, etc) on ufront.cs below.  The 03fff22c.gpubin binary is an
Intel GPU binary for vector add.  Let me know if you need anything
else.

```
ufront.cs:  /home/krentel/gtpin/
```

Dyninst master as of July 27, 2023 (8fa753d955d4), including JohnMC's
patch for arch type EM_INTELGT.

ping @hainest

**Comment by hainest** [comment]
I have a suspicion as to the problem. `DW_AT_call_line` is the line where the function is _called_ (hence, the `call` part). Could you check to see if that function is called at line 1247 in /home/johnmc/va/src/vector-add-buffers.cpp? If you follow the abstract attribute, I'm willing to venture that it leads to the correct function and declaration location.

>However 03fff22c.gpubin has two .text segments (hpcstruct-bin has only one)

Surprisingly, that is allowed by the ELF standard. Yay.

**Comment by mwkrentel** [comment]
@jmellorcrummey Do you have a copy of the `vector-add-buffers.cpp` file ?
The problem is that I copied several files off sunspot, but not that one.

**Comment by mwkrentel** [comment]
John's currently on vacation, but I poked around a little on one of
our other machines (iris.cs) and found this.  I can't be certain it's
the same `vector-add-buffers.cpp` file, but probably at least close.

I put a copy on ufront:  /home/krentel/gtpin/vector-add-buffers.cpp

Unfortunately, it only has 168 lines, so the inline function is NOT
called from line 1247.

```
$ wc vector-add-buffers.cpp 
 168  826 6205 vector-add-buffers.cpp
```

**Comment by hainest** [comment]
Hmm. I'll take a deeper look. Is it ok to copy that GPU binary off of ufront?

**Comment by jmellorcrummey** [comment]
@mwkrentel look on iris in the oneapi-examples directory for the vector-add code. i think there is a link to the particular example in iris:~johnmc/va

@hainest you can copy the gpu binary from ufront to anywhere.

**Comment by hainest** [comment]
@mwkrentel I think I might need the full source. Now that Sunspot is back up, could you pull a copy to ufront and I'll grab it from there?

**Comment by mwkrentel** [comment]
@hainest I think that `vector-add-buffers.cpp` is the only source file,
everything else comes from the compiler or the system.  However, when
we put together a database, we bundle all the files in which there
were samples taken, so I can make a tar file from that.

```
ufront:  ~krentel/gtpin/vector-src.tar.gz
```

Some of the files say `/soft/restricted/CNDA/` so let's be a little
careful with those.

**Comment by hainest** [comment]
Oh, I thought the header files were part of John's code. I have a copy of `vector-add-buffers.cpp` now. I'd recommend deleting that tarball off of ufront- I'm not entirely sure that is NDA-safe. I deleted my local copy.

**Comment by mwkrentel** [comment]
@hainest You could always take the listing from the tar file and then
look up the originals on sunspot.

**Comment by hainest** [comment]
@mwkrentel @jmellorcrummey  The last fix for this is in #1581. You'll also need https://gitlab.com/hpctoolkit/hpctoolkit/-/merge_requests/940 before you can use the new module interface this requires.

The line info for the functions declared in each module is below. Notice that everyhing in vector-add-buffers.cpp maps to line 87. That's not a Dyninst bug, the intel compiler is just doing something stupid: [readelf.lines.txt](https://github.com/dyninst/dyninst/files/12898461/readelf.lines.txt)

```
Found module '/home/johnmc/va/src/vector-add-buffers.cpp'
  Functions:
    '_entry' is at /home/johnmc/va/src/vector-add-buffers.cpp: 87
    '_ZTSZZ9VectorAddRN4sycl3_V15queueERKSt6vectorIiSaIiEES7_RS5_ENKUlRNS0_7handlerEE_clESA_EUlT_E_' is at /home/johnmc/va/src/vector-add-buffers.cpp: 87

Found module '/home/johnmc/va/src/vector-add-buffers.cpp'
  Functions:
    '_ZTSN4sycl3_V16detail18RoundedRangeKernelINS0_4itemILi1ELb1EEELi1EZZ9VectorAddRNS0_5queueERKSt6vectorIiSaIiEESB_RS9_ENKUlRNS0_7handlerEE_clESE_EUlT_E_EE' is at /soft/restricted/CNDA/updates/2023.05.15.001/oneapi/compiler/eng-20230512/compiler/linux/bin-llvm/../include/sycl/handler.hpp: 249
    '_entry' is at /soft/restricted/CNDA/updates/2023.05.15.001/oneapi/compiler/eng-20230512/compiler/linux/bin-llvm/../include/sycl/handler.hpp: 249

Found module '../gtpin/03fff22c.gpubin'
  Functions: none
```

Notice that the default module, '../gtpin/03fff22c.gpubin', contains no functions. That is because all functions found by Dyninst were successfully mapped back to a DWARF module.

**Comment by mwkrentel** [comment]
AFAIK, this was fixed in commit 356960f4a33e on Oct 13, 2023.
